[ADD] base: allow `NULLS {FIRST|LAST}` ORDER BY option

Yenthe666 opened the pull request:
Description of the issue/feature this PR addresses: Allow more dynamic SQL searches on `_order`

Current behavior before PR: You cannot order by `nulls last` since it is not required in the core from Odoo.

Desired behavior after PR is merged: You can order by `nulls last` as it is supported.
Example without `nulls last`:
```
name | score
-------------------+----------------------
abc |
def | 6.76
xxx |  5.8
```

Example with `nulls last`:
```
name | score
-------------------+----------------------
def | 6.76
xxx |  5.8
abc |
```

Fixes https://github.com/odoo/odoo/issues/51000

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

Yenthe666 commented:
@odony here is the PR to allow ordering by `nulls last`. What do you think?

odony commented on odoo/models.py:
```diff
@@ -67,7 +67,7 @@
 _schema = logging.getLogger(__name__ + '.schema')
 _unlink = logging.getLogger(__name__ + '.unlink')
 
-regex_order = re.compile('^(\s*([a-z0-9:_]+|"[a-z0-9:_]+")(\s+(desc|asc))?\s*(,|$))+(?<!,)$', re.I)
```
You want to support `nulls first` too, as it's not the default for ASC ordering.
And it's possible to specify a NULLS option even without ASC/DESC.

The following expression should do that :
`r'^(\s*([a-z0-9:_]+|"[a-z0-9:_]+")(\s+(desc|asc))?(\s+nulls\s+(first|last))?\s*(,|$))+(?<!,)$'`

Besides, the string should be a raw string as a best practice (it wasn't because we overlooked it)

odony commented:
> We could also do to allow even more options:

I don't understand what other expressions you have in mind ("ORDER BY foo NOT NULL DESC" doesn't work). Anyway, let's keep it simple.

Yenthe666 commented:
Changes, tested on multiple models/options, updated & also modified the commit message @odony.
Let me know if you'd like anything else :) @rco-odoo your feedback is welcomed too!

Yenthe666 commented:
@odony I was checking this out with @holdenrehg but we didn't really get to a fully working solution. See our sample diff at https://github.com/holdenrehg/odoo/commit/d6d38c1bc809cff4acc5bf2b1d2c8fb14bc0e35f 

For example, putting NULLS FIRST on a desc order puts NULL rows last.
It seems to always take ASC or DESC as precedence actually.

Do you know why this is the case or what we're doing wrong? Can you let us know what you think about the tests also please?

Yenthe666 commented:
@odony or @mart-e could one of you maybe give a hand? It would be nice if we can still make this land in V14 (sorry that I'm late).

Yenthe666 commented:
@odony or @mart-e just a reminder in the hope that we can still make it

Yenthe666 commented:
@odony / @mart-e any update? I'd love this so much in V14.

rco-odoo reviewed (changes requested):
This definitively needs some unit test!  I will not accept it without tests.
There are a few places where we *parse* this thing, and we don't use a dedicated parser.

We also need to cover cases where the order is *reversed* programmatically, see
https://github.com/odoo/odoo/blob/be1b2d4f4c6bd7c78cbc51e7fadeac51f38e9194/odoo/models.py#L4306-L4312
Suppose we have `order='partner_id DESC'`, and model `res.partner` has `_order = 'name, date DESC'`, the query will `ORDER BY partner.name DESC, partner.date ASC`, where `partner` is a LEFT JOIN on table `res_partner`.  Now consider `order='partner_id DESC NULLS LAST'`, what should be the query's ORDER BY clause?

Yenthe666 commented:
@rco-odoo thanks for the feedback! See my original comment about a first test suite:
>  @odony I was checking this out with @holdenrehg but we didn't really get to a fully working solution. See our sample diff at holdenrehg@d6d38c1
That got us a bit confused honestly.

Fair point about `order='partner_id DESC NULLS LAST'` and honestly I'm not sure. I guess you could say both in a way make sense but I'm not sure if there's a 'right' answer here..

3piece commented:
> You'd also need approval from @rco-odoo, and it should be clear that we don't plan to expose these technical options in the web client UI (e.g. state it explicitly in the commit message).

@Yenthe666 Thanks for the work on this. I understand it did not make it into v14, but would be great to see in v15. 

@odony, There is a real use case for this in the UI. For technical teams, they need to choose between setting values to "0" for default and not allowing null values, or showing "0's" first DESC.

This confuses end-users. It would be fantastic if there was an option in the GUI / in code to either hide non-set records in sort, or "always display null values last" option. However, you would need to have both cases, for ASC and DESC.

I'll add a request for this feature with Odoo too.

wtaferner commented:
@rco-odoo @odony 
If you agree to act and review fast I will let @ap-wtioit check on the regex (totally his area) and write the requested unit tests (he loves testing) to make that finally happen in 15.0?

odony commented:
@wtaferner this needs more work, as noted by @rco-odoo

wtaferner commented:
@odony 
Well, problems are here to be resolved, right? 😉 
As no one did resolve it yet, why not agree that you support the effort and we will at least try to make it happen with a plenty of tests of course?!
But nevertheless, if you want to do it internally, no big deal for me...just wanted to have the backup/commitment before we would start working on this.
